File: sdks/go/skopli/probe_test.go
// Probe-contract conformance, driven by the SHARED fixtures in
// golden/pricing/probe/. Each case asserts probeUsable (the same internal
// function the source lifecycle uses) agrees with the shared usable/unusable
// gold, so a facade that drifts fails against shared gold.
package skopli

import (
	"encoding/json"
	"os"
	"path/filepath"
	"testing"
)

type probeCase struct {
	Name       string          `json:"name"`
	Format     string          `json:"format"`
	Payload    json.RawMessage `json:"payload"`
	PayloadRaw *string         `json:"payloadRaw"`
	Usable     bool            `json:"usable"`
}

func TestProbeConformance(t *testing.T) {
	p := filepath.Join(repoRoot(t), "golden", "pricing", "probe", "cases.json")
	b, err := os.ReadFile(p)
	if err != nil {
		t.Fatalf("read probe cases: %v", err)
	}
	var cases []probeCase
	if err := json.Unmarshal(b, &cases); err != nil {
		t.Fatalf("decode probe cases: %v", err)
	}
	for _, kase := range cases {
		t.Run(kase.Name, func(t *testing.T) {
			// an invalid-JSON payloadRaw is fed as the raw payload bytes verbatim
			payload := kase.Payload
			if kase.PayloadRaw != nil {
				payload = json.RawMessage(*kase.PayloadRaw)
			}
			catalog := &Catalog{Source: "probe", Format: kase.Format, Payload: payload}
			if got := probeUsable(catalog); got != kase.Usable {
				t.Fatalf("probeUsable = %v, want %v", got, kase.Usable)
			}
		})
	}
}

File: sdks/csharp/Skopli.Tests/ProbeTest.cs
using System.Text.Json;
using System.Text.Json.Nodes;
using Skopli;
using Skopli.Internal;
using Xunit;

namespace Skopli.Tests;

/// <summary>
/// Drives the C# probe (<c>Probe.Usable</c>, the same internal function the
/// source lifecycle uses) against the SHARED probe-contract fixtures in
/// <c>golden/pricing/probe/</c>. A facade that drifts from the shared
/// usable/unusable gold fails here.
/// </summary>
public sealed class ProbeTest
{
    private static string RepoRoot()
    {
        string dir = AppContext.BaseDirectory;
        while (dir is not null && !Directory.Exists(Path.Combine(dir, "golden")))
        {
            dir = Path.GetDirectoryName(dir)!;
        }
        Assert.False(dir is null, "could not locate repo root (golden/)");
        return dir!;
    }

    public static IEnumerable<object[]> Cases()
    {
        string path = Path.Combine(RepoRoot(), "golden", "pricing", "probe", "cases.json");
        JsonArray cases = JsonNode.Parse(File.ReadAllText(path))!.AsArray();
        foreach (JsonNode? node in cases)
        {
            JsonObject c = node!.AsObject();
            string name = (string)c["name"]!;
            string format = (string)c["format"]!;
            bool usable = (bool)c["usable"]!;
            // an invalid-JSON payloadRaw is fed as the raw payload text verbatim
            string payload = c.ContainsKey("payloadRaw")
                ? (string)c["payloadRaw"]!
                : c["payload"]!.ToJsonString();
            yield return new object[] { name, format, payload, usable };
        }
    }

    [Theory]
    [MemberData(nameof(Cases))]
    public void ProbeConformance(string name, string format, string payload, bool usable)
    {
        var catalog = new Catalog { Source = "probe", Format = format, Payload = payload };
        Assert.True(Probe.Usable("probe", catalog) == usable, name);
    }
}
